In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import date

In [2]:
data=pd.read_csv("HRDataset.csv")

In [3]:
data.head()

Employee_Name  EmpID  MarriedID  MaritalStatusID  GenderID  \
0       Adinolfi, Wilson  K  10026          0                0         1   
1  Ait Sidi, Karthikeyan     10084          1                1         1   
2         Akinkuolie, Sarah  10196          1                1         0   
3              Alagbe,Trina  10088          1                1         0   
4          Anderson, Carol   10069          0                2         0   

   EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  Salary  ...  \
0            1       5            4                       0   62506  ...   
1            5       3            3                       0  104437  ...   
2            5       5            3                       0   64955  ...   
3            1       5            3                       0   64991  ...   
4            5       5            3                       0   50825  ...   

      ManagerName  ManagerID RecruitmentSource PerformanceScore  \
0  Michael Albert       22.0          LinkedIn          Exceeds   
1      Simon Roup        4.0            Indeed      Fully Meets   
2  Kissy Sullivan       20.0          LinkedIn      Fully Meets   
3    Elijiah Gray       16.0            Indeed      Fully Meets   
4  Webster Butler       39.0     Google Search      Fully Meets   

   EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
0              4.60               5                    0   
1              4.96               3                    6   
2              3.02               3                    0   
3              4.84               5                    0   
4              5.00               4                    0   

  LastPerformanceReview_Date DaysLateLast30 Absences  
0                  1/17/2019              0        1  
1                  2/24/2016              0       17  
2                  5/15/2012              0        3  
3                 01-03-2019              0       15  
4                 02-01-2016              0        2  

[5 rows x 36 columns]

In [4]:
data.tail()

Employee_Name  EmpID  MarriedID  MaritalStatusID  GenderID  \
306      Woodson, Jason  10135          0                0         1   
307  Ybarra, Catherine   10301          0                0         0   
308    Zamora, Jennifer  10010          0                0         0   
309         Zhou, Julia  10043          0                0         0   
310       Zima, Colleen  10271          0                4         0   

     EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  Salary  ...  \
306            1       5            3                       0   65893  ...   
307            5       5            1                       0   48513  ...   
308            1       3            4                       0  220450  ...   
309            1       3            3                       0   89292  ...   
310            1       5            3                       0   45046  ...   

        ManagerName  ManagerID  RecruitmentSource PerformanceScore  \
306  Kissy Sullivan       20.0           LinkedIn      Fully Meets   
307  Brannon Miller       12.0      Google Search              PIP   
308      Janet King        2.0  Employee Referral          Exceeds   
309      Simon Roup        4.0  Employee Referral      Fully Meets   
310   David Stanley       14.0           LinkedIn      Fully Meets   

     EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
306              4.07               4                    0   
307              3.20               2                    0   
308              4.60               5                    6   
309              5.00               3                    5   
310              4.50               5                    0   

    LastPerformanceReview_Date DaysLateLast30 Absences  
306                  2/28/2019              0       13  
307                 09-02-2015              5        4  
308                  2/21/2019              0       16  
309                 02-01-2019              0       11  
310                  1/30/2019              0        2  

[5 rows x 36 columns]

In [5]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311 entries, 0 to 310
Data columns (total 36 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Employee_Name               311 non-null    object 
 1   EmpID                       311 non-null    int64  
 2   MarriedID                   311 non-null    int64  
 3   MaritalStatusID             311 non-null    int64  
 4   GenderID                    311 non-null    int64  
 5   EmpStatusID                 311 non-null    int64  
 6   DeptID                      311 non-null    int64  
 7   PerfScoreID                 311 non-null    int64  
 8   FromDiversityJobFairID      311 non-null    int64  
 9   Salary                      311 non-null    int64  
 10  Termd                       311 non-null    int64  
 11  PositionID                  311 non-null    int64  
 12  Position                    311 non-null    object 
 13  State                       311 non

In [6]:
data.size

11196

In [7]:
data.shape

(311, 36)

In [8]:
data.isnull().sum()

Employee_Name                   0
EmpID                           0
MarriedID                       0
MaritalStatusID                 0
GenderID                        0
EmpStatusID                     0
DeptID                          0
PerfScoreID                     0
FromDiversityJobFairID          0
Salary                          0
Termd                           0
PositionID                      0
Position                        0
State                           0
Zip                             0
DOB                             0
Sex                             0
MaritalDesc                     0
CitizenDesc                     0
HispanicLatino                  0
RaceDesc                        0
DateofHire                      0
DateofTermination             207
TermReason                      0
EmploymentStatus                0
Department                      0
ManagerName                     0
ManagerID                       8
RecruitmentSource               0
PerformanceSco

Observation: Date of termination having null value, 207.

From the data we can conclude that the termination is An employee may be terminated from a job of their own free will or following a decision made by the employer.

however we can conclude that, all the employee are not involved in the stage of termination beacause they are still in the company, thats why the data is empty

In [9]:
#Treat the empty data

In [10]:
data['DateofTermination']=data['DateofTermination'].replace(to_replace=np.nan,value=0)

In [11]:
print(data['DateofTermination'].isnull().sum())

0


In [12]:
data[data['ManagerID'].isnull()]

Employee_Name  EmpID  MarriedID  MaritalStatusID  GenderID  \
19        Becker, Scott  10277          0                0         1   
30   Buccheri, Joseph    10184          0                0         1   
44    Chang, Donovan  E  10154          0                0         1   
88      Fancett, Nicole  10136          0                0         0   
135     Hutter, Rosalie  10214          0                3         0   
177   Manchester, Robyn  10077          1                1         0   
232     Rivera, Haley    10011          1                1         0   
251      Sewkumar, Nori  10071          0                0         0   

     EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  Salary  ...  \
19             3       5            3                       0   53250  ...   
30             1       5            3                       0   65288  ...   
44             1       5            3                       0   60380  ...   
88             1       5            3                       0   65902  ...   
135            2       5            3                       0   64995  ...   
177            2       5            3                       0   62385  ...   
232            1       5            4                       0   46738  ...   
251            3       5            3                       0   50923  ...   

        ManagerName  ManagerID RecruitmentSource PerformanceScore  \
19   Webster Butler        NaN          LinkedIn      Fully Meets   
30   Webster Butler        NaN     Google Search      Fully Meets   
44   Webster Butler        NaN          LinkedIn      Fully Meets   
88   Webster Butler        NaN          LinkedIn      Fully Meets   
135  Webster Butler        NaN            Indeed      Fully Meets   
177  Webster Butler        NaN          LinkedIn      Fully Meets   
232  Webster Butler        NaN     Google Search          Exceeds   
251  Webster Butler        NaN     Google Search      Fully Meets   

     EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
19               4.20               4                    0   
30               3.19               3                    0   
44               3.80               5                    0   
88               4.00               4                    0   
135              4.50               3                    0   
177              5.00               3                    0   
232              4.36               5                    0   
251              5.00               5                    0   

    LastPerformanceReview_Date DaysLateLast30 Absences  
19                  01-11-2019              0       13  
30                  02-01-2019              0        9  
44                   1/14/2019              0        4  
88                  01-07-2019              0        7  
135                  2/14/2019              0        6  
177                  1/21/2019              0        4  
232                 02-11-2019              0       16  
251                 02-06-2019              0       14  

[8 rows x 36 columns]

In [13]:
Manager_Data=data.loc[:,['Department','ManagerName','ManagerID']]

In [14]:
Manager_Data

Department     ManagerName  ManagerID
0    Production         Michael Albert       22.0
1                IT/IS      Simon Roup        4.0
2    Production         Kissy Sullivan       20.0
3    Production           Elijiah Gray       16.0
4    Production         Webster Butler       39.0
..                 ...             ...        ...
306  Production         Kissy Sullivan       20.0
307  Production         Brannon Miller       12.0
308              IT/IS      Janet King        2.0
309              IT/IS      Simon Roup        4.0
310  Production          David Stanley       14.0

[311 rows x 3 columns]

In [15]:
Manager_Data.tail()

Department     ManagerName  ManagerID
306  Production         Kissy Sullivan       20.0
307  Production         Brannon Miller       12.0
308              IT/IS      Janet King        2.0
309              IT/IS      Simon Roup        4.0
310  Production          David Stanley       14.0

In [16]:
n_Value=Manager_Data[Manager_Data['ManagerID'].isnull()]
n_Value

Department     ManagerName  ManagerID
19   Production         Webster Butler        NaN
30   Production         Webster Butler        NaN
44   Production         Webster Butler        NaN
88   Production         Webster Butler        NaN
135  Production         Webster Butler        NaN
177  Production         Webster Butler        NaN
232  Production         Webster Butler        NaN
251  Production         Webster Butler        NaN

The empty manager id from production department. 

We need to fetch the details of depatment "production".

In [17]:
Manager_Data

Department     ManagerName  ManagerID
0    Production         Michael Albert       22.0
1                IT/IS      Simon Roup        4.0
2    Production         Kissy Sullivan       20.0
3    Production           Elijiah Gray       16.0
4    Production         Webster Butler       39.0
..                 ...             ...        ...
306  Production         Kissy Sullivan       20.0
307  Production         Brannon Miller       12.0
308              IT/IS      Janet King        2.0
309              IT/IS      Simon Roup        4.0
310  Production          David Stanley       14.0

[311 rows x 3 columns]

In [18]:
Manager_Data.loc[Manager_Data['ManagerName'] == 'Webster Butler']

Department     ManagerName  ManagerID
4    Production         Webster Butler       39.0
19   Production         Webster Butler        NaN
30   Production         Webster Butler        NaN
44   Production         Webster Butler        NaN
65   Production         Webster Butler       39.0
88   Production         Webster Butler        NaN
89   Production         Webster Butler       39.0
105  Production         Webster Butler       39.0
124  Production         Webster Butler       39.0
135  Production         Webster Butler        NaN
151  Production         Webster Butler       39.0
174  Production         Webster Butler       39.0
177  Production         Webster Butler        NaN
198  Production         Webster Butler       39.0
206  Production         Webster Butler       39.0
214  Production         Webster Butler       39.0
232  Production         Webster Butler        NaN
251  Production         Webster Butler        NaN
276  Production         Webster Butler       39.0
280  Production         Webster Butler       39.0
300  Production         Webster Butler       39.0

In [19]:
data['ManagerID']=data['ManagerID'].replace(to_replace=np.nan,value=39.0)

In [20]:
data.isnull().sum()

Employee_Name                 0
EmpID                         0
MarriedID                     0
MaritalStatusID               0
GenderID                      0
EmpStatusID                   0
DeptID                        0
PerfScoreID                   0
FromDiversityJobFairID        0
Salary                        0
Termd                         0
PositionID                    0
Position                      0
State                         0
Zip                           0
DOB                           0
Sex                           0
MaritalDesc                   0
CitizenDesc                   0
HispanicLatino                0
RaceDesc                      0
DateofHire                    0
DateofTermination             0
TermReason                    0
EmploymentStatus              0
Department                    0
ManagerName                   0
ManagerID                     0
RecruitmentSource             0
PerformanceScore              0
EngagementSurvey              0
EmpSatis

# Data Analysis 

1) Count of employees in each department

In [23]:
Count_Emp=data.groupby('Department')['Department'].count()

In [24]:
Count_Emp

Department
Admin Offices             9
Executive Office          1
IT/IS                    50
Production              209
Sales                    31
Software Engineering     11
Name: Department, dtype: int64

2) Average salary of employess in each depatment

In [25]:
Avg_Salary=data.groupby('Department')['Salary'].mean()

In [26]:
Avg_Salary.round(2)

Department
Admin Offices            71791.89
Executive Office        250000.00
IT/IS                    97064.64
Production               59953.55
Sales                    69061.26
Software Engineering     94989.45
Name: Salary, dtype: float64

3) Calculate the average of salary of employees based on gender.

In [27]:
sex_salary=data.groupby(['Department','Sex'],as_index=False).Salary.mean()

In [28]:
sex_salary

Department Sex         Salary
0          Admin Offices   F   70926.333333
1          Admin Offices  M    73523.000000
2       Executive Office   F  250000.000000
3                  IT/IS   F   94593.000000
4                  IT/IS  M    99006.642857
5      Production          F   59620.420635
6      Production         M    60459.253012
7                  Sales   F   72016.666667
8                  Sales  M    66290.562500
9   Software Engineering   F   96906.166667
10  Software Engineering  M    92689.400000

4) Most popular RecruitmentSource of the company based on department.

In [33]:
Recruitmentsource=data.groupby(['Department','RecruitmentSource']).size().reset_index(name='counts')

In [34]:
Recruitmentsource

Department        RecruitmentSource  counts
0          Admin Offices            CareerBuilder       1
1          Admin Offices       Diversity Job Fair       3
2          Admin Offices                   Indeed       2
3          Admin Offices                    Other       1
4          Admin Offices                  Website       2
5       Executive Office                   Indeed       1
6                  IT/IS            CareerBuilder       2
7                  IT/IS       Diversity Job Fair       4
8                  IT/IS        Employee Referral      13
9                  IT/IS            Google Search       1
10                 IT/IS                   Indeed      19
11                 IT/IS                 LinkedIn      11
12     Production                   CareerBuilder      16
13     Production              Diversity Job Fair      17
14     Production               Employee Referral      16
15     Production                   Google Search      44
16     Production                          Indeed      49
17     Production                        LinkedIn      61
18     Production         On-line Web application       1
19     Production                           Other       1
20     Production                         Website       4
21                 Sales            CareerBuilder       3
22                 Sales       Diversity Job Fair       2
23                 Sales        Employee Referral       2
24                 Sales            Google Search       2
25                 Sales                   Indeed      13
26                 Sales                 LinkedIn       2
27                 Sales                  Website       7
28  Software Engineering            CareerBuilder       1
29  Software Engineering       Diversity Job Fair       3
30  Software Engineering            Google Search       2
31  Software Engineering                   Indeed       3
32  Software Engineering                 LinkedIn       2

5) Total Number of working days for terminated employee.

In [35]:
Employee_Name=input("Enter the Employee name:")

Enter the Employee name:Hari


In [36]:
Start_year=int(input("Enter the start year: "))
Start_month=int(input("Enter the start month: "))
Start_day=int(input("Enter the start day: "))
Start_Date=date(Start_year,Start_month,Start_day)
Start_Date

Enter the start year: 2011
Enter the start month: 5
Enter the start day: 3


datetime.date(2011, 5, 3)

In [37]:
T_year=int(input("Enter the Terminated year: "))
T_month=int(input("Enter the Terminated month: "))
T_day=int(input("Enter the Terminated day: "))
T_Date=date(T_year,T_month,T_day)

Enter the Terminated year: 2013
Enter the Terminated month: 6
Enter the Terminated day: 6


In [38]:
business_days=pd.bdate_range(Start_Date,T_Date)

In [39]:
business_days

DatetimeIndex(['2011-05-03', '2011-05-04', '2011-05-05', '2011-05-06',
               '2011-05-09', '2011-05-10', '2011-05-11', '2011-05-12',
               '2011-05-13', '2011-05-16',
               ...
               '2013-05-24', '2013-05-27', '2013-05-28', '2013-05-29',
               '2013-05-30', '2013-05-31', '2013-06-03', '2013-06-04',
               '2013-06-05', '2013-06-06'],
              dtype='datetime64[ns]', length=548, freq='B')

In [40]:
print("Employee name:",Employee_Name,"Working days",len(business_days))

Employee name: Hari Working days 548


6) Fetch the manager name of departments.

In [41]:
Manager_Details=data.groupby(['Department','ManagerName','ManagerID']).size().reset_index(name='counts')

In [42]:
Manager_Details.set_index("ManagerName",inplace=True)

In [43]:
Manager_Details

Department  ManagerID  counts
ManagerName                                                
Board of Directors         Admin Offices        9.0       1
Brandon R. LeBlanc         Admin Offices        1.0       6
Brandon R. LeBlanc         Admin Offices        3.0       1
Janet King                 Admin Offices        2.0       1
Board of Directors      Executive Office        9.0       1
Brian Champaigne                   IT/IS       13.0       8
Eric Dougall                       IT/IS        6.0       4
Janet King                         IT/IS        2.0       1
Jennifer Zamora                    IT/IS        5.0       6
Peter Monroe                       IT/IS        7.0      14
Simon Roup                         IT/IS        4.0      17
Amy Dunn               Production              11.0      21
Brannon Miller         Production              12.0      22
David Stanley          Production              14.0      21
Elijiah Gray           Production              16.0      22
Janet King             Production               2.0      15
Kelley Spirea          Production              18.0      22
Ketsia Liebig          Production              19.0      21
Kissy Sullivan         Production              20.0      22
Michael Albert         Production              22.0      21
Michael Albert         Production              30.0       1
Webster Butler         Production              39.0      21
Debra Houlihan                     Sales       15.0       3
Janet King                         Sales        2.0       1
John Smith                         Sales       17.0      14
Lynn Daneault                      Sales       21.0      13
Alex Sweetwater     Software Engineering       10.0       9
Janet King          Software Engineering        2.0       1
Jennifer Zamora     Software Engineering        5.0       1

In [56]:

Manager_Name=input("Enter the manager Name: ")
Dep_value=Manager_Details.loc[[Manager_Name],['Department']]


Enter the manager Name: Lynn Daneault


In [57]:
Dep_value

Department
ManagerName             
Lynn Daneault      Sales

7) Fetch Employee name and number of absences.

In [47]:
Emp_Attendance=data.loc[:,['Department','ManagerName','Employee_Name','Absences']]

In [48]:
Emp_Attendance

Department     ManagerName             Employee_Name  Absences
0    Production         Michael Albert       Adinolfi, Wilson  K         1
1                IT/IS      Simon Roup  Ait Sidi, Karthikeyan           17
2    Production         Kissy Sullivan         Akinkuolie, Sarah         3
3    Production           Elijiah Gray              Alagbe,Trina        15
4    Production         Webster Butler          Anderson, Carol          2
..                 ...             ...                       ...       ...
306  Production         Kissy Sullivan            Woodson, Jason        13
307  Production         Brannon Miller        Ybarra, Catherine          4
308              IT/IS      Janet King          Zamora, Jennifer        16
309              IT/IS      Simon Roup               Zhou, Julia        11
310  Production          David Stanley             Zima, Colleen         2

[311 rows x 4 columns]

Conclusion: HR data Analysis using python